src/oaep.rs

```diff
@@ -1,20 +1,169 @@
+    /// If a label is needed use `PaddingScheme::new_oaep_with_label` or `PaddingScheme::new_oaep_with_mgf_hash_with_label`.
+    ///
+    /// # Example
+    /// ```
+    /// use sha1::Sha1;
+    /// use sha2::Sha256;
+    /// use rsa::{BigUint, RsaPublicKey, Oaep, PublicKey};
+    /// use base64ct::{Base64, Encoding};
+    ///
+    /// let n = Base64::decode_vec("ALHgDoZmBQIx+jTmgeeHW6KsPOrj11f6CvWsiRleJlQpW77AwSZhd21ZDmlTKfaIHBSUxRUsuYNh7E2SHx8rkFVCQA2/gXkZ5GK2IUbzSTio9qXA25MWHvVxjMfKSL8ZAxZyKbrG94FLLszFAFOaiLLY8ECs7g+dXOriYtBwLUJK+lppbd+El+8ZA/zH0bk7vbqph5pIoiWggxwdq3mEz4LnrUln7r6dagSQzYErKewY8GADVpXcq5mfHC1xF2DFBub7bFjMVM5fHq7RK+pG5xjNDiYITbhLYrbVv3X0z75OvN0dY49ITWjM7xyvMWJXVJS7sJlgmCCL6RwWgP8PhcE=").unwrap();
+    /// let e = Base64::decode_vec("AQAB").unwrap();
+    ///
+    /// let mut rng = rand::thread_rng();
+    /// let key = RsaPublicKey::new(BigUint::from_bytes_be(&n), BigUint::from_bytes_be(&e)).unwrap();
+    /// let padding = Oaep::new_with_mgf_hash::<Sha256, Sha1>();
+    /// let encrypted_data = key.encrypt(&mut rng, padding, b"secret").unwrap();
+    /// ```
+    pub fn new_with_mgf_hash<
+        T: 'static + Digest + DynDigest + Send + Sync,
+        U: 'static + Digest + DynDigest + Send + Sync,
+    >() -> Self {
+        Self {
+            digest: Box::new(T::new()),
+            mgf_digest: Box::new(U::new()),
+            label: None,
+        }
+    }
+
+    /// Create a new OAEP `PaddingScheme` with an associated `label`, using `T` as the hash function for the label, and `U` as the hash function for MGF1.
+    pub fn new_with_mgf_hash_and_label<
+        T: 'static + Digest + DynDigest + Send + Sync,
+        U: 'static + Digest + DynDigest + Send + Sync,
+        S: AsRef<str>,
+    >(
+        label: S,
+    ) -> Self {
+        Self {
+            digest: Box::new(T::new()),
+            mgf_digest: Box::new(U::new()),
+            label: Some(label.as_ref().to_string()),
+        }
+    }
+}
+
+impl PaddingScheme for Oaep {
+    fn decrypt(
+        mut self,
+        rng: Option<&mut impl CryptoRngCore>,
+        priv_key: &impl PrivateKey,
+        ciphertext: &[u8],
+    ) -> Result<Vec<u8>> {
+        decrypt(
+            rng,
+            priv_key,
+            ciphertext,
+            &mut *self.digest,
+            &mut *self.mgf_digest,
+            self.label,
+        )
+    }
+
+    fn encrypt(
+        mut self,
+        rng: &mut impl CryptoRngCore,
+        pub_key: &impl PublicKey,
+        msg: &[u8],
+    ) -> Result<Vec<u8>> {
+        encrypt(
+            rng,
+            pub_key,
+            msg,
+            &mut *self.digest,
+            &mut *self.mgf_digest,
+            self.label,
+        )
+    }
+}
+
+impl fmt::Debug for Oaep {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("OAEP")
+            .field("label", &self.label)
+            .finish_non_exhaustive()
```

Comment:
this is a bit odd, not listing all fields and having the struct not actually be non exhaustive, while printing it as such

Reply:
Well, the old impl looked like this...

```rust
            PaddingScheme::OAEP { ref label, .. } => {
                // TODO: How to print the digest name?
                write!(f, "PaddingScheme::OAEP({:?})", label)
            }
```

It's at least a step up from that. I suppose it could list out the other field names, but we have nothing to display for them.

Reply:
Yeah, previous one wasn't great either. What I usually do is add the fields and print the type description or sth like that, so that at least there is some info on what is there.

Reply:
[69c9ba0](https://github.com/RustCrypto/RSA/pull/244/commits/69c9ba0e91a0684fb0f2d38d364e09d617097cee) adds the field names.

Perhaps we could add a `DynDigestWithOid` trait to `Digest` which would make it possible to get the OID for the digest, but for now I just used `...`